crates/apub/src/api/user_settings_backup.rs

```diff
@@ -0,0 +1,376 @@
+    banner: local_user_view.person.banner,
+    matrix_id: local_user_view.person.matrix_user_id,
+    bot_account: local_user_view.person.bot_account.into(),
+    settings: Some(local_user_view.local_user),
+    followed_communities: vec_into(lists.followed_communities),
+    blocked_communities: vec_into(lists.blocked_communities),
+    blocked_users: lists.blocked_users.into_iter().map(Into::into).collect(),
+    saved_posts: lists.saved_posts.into_iter().map(Into::into).collect(),
+    saved_comments: lists.saved_comments.into_iter().map(Into::into).collect(),
+  }))
+}
+
+#[tracing::instrument(skip(context))]
+pub async fn import_settings(
+  data: Json<UserSettingsBackup>,
+  local_user_view: LocalUserView,
+  context: Data<LemmyContext>,
+) -> Result<Json<SuccessResponse>, LemmyError> {
+  let display_name = Some(sanitize_html_api_opt(&data.display_name));
+  let bio = Some(sanitize_html_api_opt(&data.bio));
+
+  let person_form = PersonUpdateForm {
+    display_name,
+    bio,
+    matrix_user_id: Some(data.matrix_id.clone()),
+    bot_account: data.bot_account,
+    ..Default::default()
+  };
+  Person::update(&mut context.pool(), local_user_view.person.id, &person_form).await?;
+
+  let local_user_form = LocalUserUpdateForm {
+    show_nsfw: data.settings.as_ref().map(|s| s.show_nsfw),
+    theme: data.settings.as_ref().map(|s| s.theme.clone()),
+    default_sort_type: data.settings.as_ref().map(|s| s.default_sort_type),
+    default_listing_type: data.settings.as_ref().map(|s| s.default_listing_type),
+    interface_language: data.settings.as_ref().map(|s| s.interface_language.clone()),
+    show_avatars: data.settings.as_ref().map(|s| s.show_avatars),
+    send_notifications_to_email: data
+      .settings
+      .as_ref()
+      .map(|s| s.send_notifications_to_email),
+    show_scores: data.settings.as_ref().map(|s| s.show_scores),
+    show_bot_accounts: data.settings.as_ref().map(|s| s.show_bot_accounts),
+    show_read_posts: data.settings.as_ref().map(|s| s.show_read_posts),
+    open_links_in_new_tab: data.settings.as_ref().map(|s| s.open_links_in_new_tab),
+    blur_nsfw: data.settings.as_ref().map(|s| s.blur_nsfw),
+    auto_expand: data.settings.as_ref().map(|s| s.auto_expand),
+    infinite_scroll_enabled: data.settings.as_ref().map(|s| s.infinite_scroll_enabled),
+    post_listing_mode: data.settings.as_ref().map(|s| s.post_listing_mode),
+    ..Default::default()
+  };
+  LocalUser::update(
+    &mut context.pool(),
+    local_user_view.local_user.id,
+    &local_user_form,
+  )
+  .await?;
+
+  let url_count = data.followed_communities.len()
+    + data.blocked_communities.len()
+    + data.blocked_users.len()
+    + data.saved_posts.len()
+    + data.saved_comments.len();
+  if url_count > MAX_URL_IMPORT_COUNT {
+    Err(LemmyErrorType::UserBackupTooLarge)?;
+  }
+
+  spawn_try_task(async move {
+    let person_id = local_user_view.person.id;
+    try_join_all(data.followed_communities.iter().map(|followed| async {
```

Comment:
i don't think try_join_all is a good method to use here for two reasons:

1. if one of the follows fail, the others should still go through. for example if a user has followed communities on instances that no longer exist then the import will fail at a random point and leave the import partially done. (from the docs: If any future returns an error then all other futures will be canceled and an error will be returned immediately)

    Instead, a method should be used where each error is handled separately with the main import keeping going.
3. There should be a limit on the concurrency. If the user follows 500 communities we don't want the server to fetch and subscribe to all of those 100 communities at once because rate limits might cause that to fail. It should either be fully sequential or with a fixed limit on concurrency. 

    This should be possible by converting the iterator into a stream and calling `.buffered(10)` (see https://stackoverflow.com/a/70871743/2639190 )

Reply:
If these links are to be believed then join_all runs futures concurrently, not in parallel. Though I havent seen a clear confirmation for that in the docs.
- https://stackoverflow.com/questions/63756169/rust-futuresfuturejoin-all-await-runs-futures-sequentially-but-parall
- https://www.reddit.com/r/rust/comments/dt6u0s/joined_futures_will_run_concurrently_not_in/

If this is correct it would be enough to switch from `try_join_all` to `join_all` and collect the errors manually. I also tried to switch to the method you suggested, but its not working because the compiler throws an error about `context` geting moved.

```

    futures::stream::iter(data.followed_communities.clone().into_iter().map(
      |followed| async move {
        // need to reset outgoing request count to avoid running into limit
        let context_ = context_.reset_request_count();
        let community = followed.dereference(&context_).await?;
        let form = CommunityFollowerForm {
          person_id,
          community_id: community.id,
          pending: true,
        };
        CommunityFollower::follow(&mut context_.pool(), &form).await?;
        LemmyResult::Ok(())
      },
    ))
    .buffer_unordered(10)
    .collect::<Vec<_>>()
    .await;
```

Reply:
Okay I added error handling for remote fetches now. Also tried to return a list of failed items in the api response but couldnt get that working. For now its only being logged.

Reply:
> concurrently, not in parallel

Makes sense, but that'll still mean that it will send out 100 requests at almost the same time (since it'll mainly be waiting on the HTTP I/O), which will potentially cause issues

Reply:
Got it working now. The code is rather verbose and repetetive now. It could potentially be simplified by moving the iter logic into a helper function, but that gets very complex with lots of generics and lambdas.